Photo to src

zeroadam opened the pull request:
Move Photo class to Friendica\Object namespace.
Imagick gave me a fit, but I'm no longer getting errors.

Related to #3878

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -0,0 +1,1206 @@
+	private function loadData($data)
+	{
+		if ($this->isImagick()) {
+			$this->image = new \Imagick();
```
Why not add a couple of `use` statements? Wouldn't it remove the need for the horrible syntax like `$this->image->\Imagick::readImageBlob($data);`?

zeroadam replied:
Can you give me an example of how? I am having trouble getting this to work...

MrPetovan replied:
Just add 
```php
use Imagick;
use ImagickPixel;
```
After the other `use` statements, and you don't have to add the backslashes in front of the `Imagick` class names. Not sure about the individual method calls.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -0,0 +1,1206 @@
+	 *
+	 * @return array Returns array of the different avatar sizes
+	 */
+	public function updateContactAvatar($avatar, $uid, $cid, $force = false)
```
Can you move this to `\Friendica\Object\Contact::updateAvatar()` please?

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -0,0 +1,1206 @@
+			$micro = System::baseUrl() . '/images/person-48.jpg';
+		}
+	
+		return(array($photo, $thumb, $micro));
```
Standards: Please remove parentheses from language constructs.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -0,0 +1,1206 @@
+	 * @param string $url url
+	 * @return object
+	 */
+	public function getPhotoInfo($url)
```
Please rename to `getInfoFromURL()`.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -0,0 +1,1206 @@
+	}
+
+	/**
+	 * @brief This function doesn't seem to be used
```
If the function isn't used anywhere, please scrap it. We will be able to find it again in the git history if needed. There already is a `Photo::store()` function anyway.

annando replied:
It is used in the addon "fromgplus".

MrPetovan replied:
What is the difference with `Photo::store()` ?

annando replied:
Good question. I guess we have to check the differences by comparing the functionality.

zeroadam commented:
Phew! I think that did it... I don't know why I was having so much trouble with Imagick I had tried use statement early on but was getting errors, I think now it must have been related to something else. Should be all cleaned up now, and I'm back to no errors!

annando commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,244 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
-	return array("width" => $dest_width, "height" => $dest_height);
-}
-
-function store_photo(App $a, $uid, $imagedata = "", $url = "") {
```
You removed this function, but you haven't changed the corresponding calls to this function in the file ```fromgplus.php```. Please make sure that either the call is changed or this function is restored. And no: I don't know the difference between the two functions that store photos.

zeroadam commented:
Replaced the function and updated calls from addon. It looks like it was doing more than just storing the photo like in the store() function. I also fixed an error I got about using $this instead of self:: so If someone could look again and see if there are more of these that should be replaced, just in case!

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
+				$image->setInterlaceScheme(Imagick::INTERLACE_PLANE);
+			} else {
+				$ext = pathinfo($filename, PATHINFO_EXTENSION);
+				$types = $this->supportedTypes();
```
This should be static as well.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
+	 *
+	 * @return object
+	 */
+	public function guessImageType($filename, $fromcurl = false)
```
This function should be static.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
+	 * @param boolean $quit_on_error optional, default false
+	 * @return array
+	 */
+	public function importProfilePhoto($photo, $uid, $cid, $quit_on_error = false)
```
This function should be static.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
+	 * @param string $url url
+	 * @return object
+	 */
+	public function getInfoFromURL($url)
```
This function should be static.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
+	 * @param integer $max    max
+	 * @return array
+	 */
+	public function scaleImageTo($width, $height, $max)
```
This function should be static.

MrPetovan commented on src/Object/Photo.php:
```diff
@@ -653,417 +758,408 @@ public function store($uid, $cid, $rid, $filename, $album, $scale, $profile = 0,
+	 * @param string  $url       optional, default empty
+	 * @return array
+	 */
+	private function storePhoto(App $a, $uid, $imagedata = "", $url = "")
```
This function should be public and static.

annando commented:
I'm now getting errors like this:
```
PHP Fatal error:  Uncaught ImagickException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/353 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php:134
Stack trace:
#0 /.../src/Object/Photo.php(134): Imagick->readimageblob('<!DOCTYPE html\n...')
#1 /.../src/Object/Photo.php(76): Friendica\Object\Photo->loadData('<!DOCTYPE html\n...')
#2 /.../src/Object/Photo.php(843): Friendica\Object\Photo->__construct('<!DOCTYPE html\n...', 'image/jpeg')
#3 /.../src/Object/Contact.php(845): Friendica\Object\Photo::importProfilePhoto('https://gnusoci...', '1', '11355', true)
#4 /.../src/Protocol/OStatus.php(206): Friendica\Object\Contact::updateAvatar('https://gnusoci...', '1', '11355')
#5 /.../src/Protocol/OStatus.php(388): Friendica\Protocol\OStatus::fetchAuthor(Object(DOMXPath), NULL, Array, Array, false)
#6 /.../src/Object/Photo.php on line 134
PHP Fatal error:  Uncaught ImagickException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/353 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php:134
Stack trace:
#0 /.../src/Object/Photo.php(134): Imagick->readimageblob('<!DOCTYPE html\n...')
#1 /.../src/Object/Photo.php(76): Friendica\Object\Photo->loadData('<!DOCTYPE html\n...')
#2 /.../src/Object/Photo.php(843): Friendica\Object\Photo->__construct('<!DOCTYPE html\n...', 'image/jpeg')
#3 /.../src/Object/Contact.php(845): Friendica\Object\Photo::importProfilePhoto('https://gnusoci...', 0, '9914', true)
#4 /.../src/Object/Contact.php(649): Friendica\Object\Contact::updateAvatar('https://gnusoci...', 0, '9914')
#5 /.../src/Protocol/OStatus.php(210): Friendica\Object\Contact::getIdForURL('https://gnusoci...', 0)
#6 /.../src/Protocol/OStatus.php(38 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php on line 134
PHP Fatal error:  Uncaught ImagickException: no decode delegate for this image format `' @ error/blob.c/BlobToImage/353 in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php:134
Stack trace:
#0 /.../src/Object/Photo.php(134): Imagick->readimageblob('<!DOCTYPE html\n...')
#1 /.../src/Object/Photo.php(76): Friendica\Object\Photo->loadData('<!DOCTYPE html\n...')
#2 /.../src/Object/Photo.php(843): Friendica\Object\Photo->__construct('<!DOCTYPE html\n...', 'image/jpeg')
#3 /.../src/Object/Contact.php(845): Friendica\Object\Photo::importProfilePhoto('https://indy.im...', 0, '499513', true)
#4 /.../src/Object/Contact.php(649): Friendica\Object\Contact::updateAvatar('https://indy.im...', 0, '499513')
#5 /.../src/Protocol/OStatus.php(210): Friendica\Object\Contact::getIdForURL('https://gnusoci...', 0)
#6 /.../src/Protocol/OStatus.ph in /var/www/virtual/pirati.ca/htdocs/src/Object/Photo.php on line 134
```

zeroadam commented:
This looks like html data being sent in?

MrPetovan commented:
Yep, the gnusocial URL returns plain HTML when it expects an image.

annando commented:
The problem is not that the data is bad - but that this here doesn't work: https://github.com/friendica/friendica/blob/develop/src/Object/Photo.php#L133-L138

```try``` ... ```catch``` should catch this exception, shouldn't it?

MrPetovan commented:
It’s a namespace issue then, `\ImagickException` extends `\Exception`, not
`Friendica\Object\Exception`, we should either add `use Exception;` or use
`\Exception`
On Fri, Dec 1, 2017 at 18:33 Michael Vogel <notifications@github.com> wrote:

> The problem is not that the data is bad - but that this here doesn't work:
> https://github.com/friendica/friendica/blob/develop/src/Object/Photo.php#L133-L138
>
> try ... catch should catch this exception, shouldn't it?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/friendica/friendica/pull/3982#issuecomment-348642184>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4e52Eoa6WUgSvvKbgo1xkxmVD47kjhks5s8Iy3gaJpZM4Qvh3O>
> .
>

zeroadam commented:
https://github.com/friendica/friendica/pull/3992

MrPetovan commented:
👌👍

annando commented:
I guess that we will have this issue at other places in the code as well.
